sdks/python/apache_beam/runners/portability/fn_api_runner.py

```diff
@@ -554,62 +527,277 @@ def run_stages(self,
-                                  worker_handler,  # type: WorkerHandler
-                                  context,  # type: pipeline_context.PipelineContext
-                                  pipeline_components,  # type: beam_runner_api_pb2.Components
-                                  data_side_input,  # type: DataSideInput
-                                  pcoll_buffers,  # type: Mapping[bytes, _ListBuffer]
-                                  safe_coders):
-    for (transform_id, tag), (buffer_id, si) in data_side_input.items():
-      _, pcoll_id = split_buffer_id(buffer_id)
-      value_coder = context.coders[safe_coders[
-          pipeline_components.pcollections[pcoll_id].coder_id]]
-      elements_by_window = _WindowGroupingBuffer(si, value_coder)
-      for element_data in pcoll_buffers[buffer_id]:
-        elements_by_window.append(element_data)
-
-      if si.urn == common_urns.side_inputs.ITERABLE.urn:
-        for _, window, elements_data in elements_by_window.encoded_items():
-          state_key = beam_fn_api_pb2.StateKey(
-              iterable_side_input=beam_fn_api_pb2.StateKey.IterableSideInput(
-                  transform_id=transform_id,
-                  side_input_id=tag,
-                  window=window))
-          worker_handler.state.append_raw(state_key, elements_data)
-      elif si.urn == common_urns.side_inputs.MULTIMAP.urn:
-        for key, window, elements_data in elements_by_window.encoded_items():
-          state_key = beam_fn_api_pb2.StateKey(
-              multimap_side_input=beam_fn_api_pb2.StateKey.MultimapSideInput(
-                  transform_id=transform_id,
-                  side_input_id=tag,
-                  window=window,
-                  key=key))
-          worker_handler.state.append_raw(state_key, elements_data)
+  @staticmethod
+  def _merge_metrics_from_stage_execution(
+      monitoring_infos_by_stage,  # type: Dict[str, List[Any]]
+      metrics_by_stage,  # type: Dict[str, List[beam_fn_api_pb2.Metrics]]
+      stage_results,  # type: beam_fn_api_pb2.InstructionResponse
+      stage_name  # type: str
+  ):
+    # type: (...) -> None
+    stage_minfos = stage_results.process_bundle.monitoring_infos
+    if stage_name in monitoring_infos_by_stage:
+      stage_minfos = monitoring_infos.consolidate(
+          itertools.chain(stage_minfos, monitoring_infos_by_stage[stage_name]))
+
+    monitoring_infos_by_stage[stage_name] = stage_minfos
+
+    # TODO(pabloem): Note that these are not merged.
+    metrics_by_stage[stage_name] = stage_results.process_bundle.metrics
+
+  @staticmethod
+  def _schedule_newly_ready_bundles(execution_context, input_queue_manager):
+    # type: (PipelineExecutionContext, _ProcessingQueueManager) -> None
+    """Inspect watermark pending bundles, and schedule ready ones.
+
+    This function should be called after the watermarks have been recalculated.
+    """
+    requed_inputs = []
+    while len(input_queue_manager.watermark_pending_inputs) > 0:
```

Comment:
Would this better be expressed as a for loop over all pending inputs? (Alternatively, would it make sense for the data structure to be `Dict[StageName, Heap[Tuple[Watermark, Inputs]]]` and one could then pop all (ordered) ready inputs on each watermark advance? Inspecting every pending queue at every stage execution seems like it'd be quadratic time.